content/en/blog/_posts/2023-03-30-kubescape-validating-admission-policy-library.md

```diff
@@ -0,0 +1,235 @@
+---
+layout: blog
+title: "Kubernetes Validating Admission Policies: A Practical Example"
+date: 2023-03-30T00:00:00+0000
+slug: kubescape-validating-admission-policy-library
+---
+
+**Authors**: Craig Box (ARMO), Ben Hirschberg (ARMO)
+
+Admission control is an important part of the Kubernetes control plane, with several internal
+features depending on the ability to approve or change an API object as it is submitted to the
+server. It is also useful for an administrator to be able to define business logic, or policies,
+regarding what objects can be admitted into a cluster. To better support that use case, [Kubernetes
+introduced external admission control in
+v1.7](https://kubernetes.io/blog/2017/06/kubernetes-1-7-security-hardening-stateful-application-extensibility-updates/).
+
+In addition to countless custom, internal implementations, many open source projects and commercial
+solutions implement admission controllers with user-specified policy, including
+[Kyverno](https://github.com/kyverno/kyverno) and Open Policy Agent’s
+[Gatekeeper](https://github.com/open-policy-agent/gatekeeper).
+
+While admission controllers for policy have seen adoption, there are blockers for their widespread
+use. Webhook infrastructure must be maintained as a production service, with all that entails. The
+failure case of an admission control webhook must either be closed, reducing the availability of the
+cluster; or open, negating the use of the feature for policy enforcement. The network hop and
+evaluation time makes admission control a notable component of latency when dealing with, for
+example, pods being spun up to respond to a network request in a "serverless" environment.
+
+## Validating admission policies and the Common Expression Language
+
+Version 1.26 of Kubernetes introduced, in alpha, a compromise solution. [Validating admission
+policies](/docs/reference/access-authn-authz/validating-admission-policy/) are a declarative,
+in-process alternative to admission webhooks. They use the [Common Expression
+Language](https://github.com/google/cel-spec) (CEL) to declare validation rules.
+
+CEL was developed by Google for security and policy use cases, based on learnings from the Firebase
+real-time database. Its design allows it to be safely embedded into applications and executed in
+microseconds, with limited compute and memory impact. [Validation rules for
+CRDs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
```

Comment:
Sorry, this is a remnant of writing in Google Docs and converting to Markdown.
Do you have a lint process on the docs at all?  [Istio's linter](https://github.com/istio/istio.io/blob/master/scripts/lint_site.sh) would have stopped me from doing this, so I'm not in the habit of looking.  I'll raise a FR if this would be useful/worthwhile, but it's hard to override if there are cases where it is not feasible.

Reply:
We don't have a linter, and this is a nice-to-have not a merge blocker.

Reply:
If you run the following command:

```
./scripts/linkchecker.py
```

You will receive some errors like `https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules: Should use relative paths`.

The script works for localization teams as well:

```
./scripts/linkchecker.py -f 'content/zh-cn/docs/**/*.md'
```

You may get some more complaints like `    /docs/tasks/debug/debug-cluster/crictl/: Localized page detected, please append '/zh-cn' to the target`.